src/rust/engine/process_execution/src/speculate.rs

```diff
@@ -0,0 +1,214 @@
+use std::time::{Duration, Instant};
+use std::sync::Arc;
+use tokio_timer::Delay;
+use futures::future::{Either, Future, result};
+use boxfuture::{BoxFuture, Boxable};
+use workunit_store::WorkUnitStore;
+use super::{
+  CommandRunner,
+  ExecuteProcessRequest,
+  FallibleExecuteProcessResult
+};
+
+
+#[derive(Clone)]
+pub struct SpeculatingCommandRunner {
+  primary: Arc<dyn CommandRunner>,
+  secondary: Arc<dyn CommandRunner>,
+  speculation_timeout_ms: u64,
+}
+
+impl SpeculatingCommandRunner {
+
+  pub fn new(primary: Box<dyn CommandRunner>, secondary: Box<dyn CommandRunner>, speculation_timeout_ms: u64) -> SpeculatingCommandRunner {
+    SpeculatingCommandRunner {
+      primary: primary.into(),
+      secondary: secondary.into(),
+      speculation_timeout_ms: speculation_timeout_ms
+    }
+  }
+}
+
+impl CommandRunner for SpeculatingCommandRunner {
+  fn run(
+    &self, req: ExecuteProcessRequest, workunit_store: WorkUnitStore
+  ) -> BoxFuture<FallibleExecuteProcessResult, String> {
+    let command_runner = self.clone();
+    let workunit_store_clone = workunit_store.clone();
+    let req_2 = req.clone();
+    let delay = Delay::new(Instant::now() + Duration::from_millis(self.speculation_timeout_ms));
+    self.primary.run(req, workunit_store).select2(delay).then(
```

Comment:
Assuming we need to support cancellation either way (and I don't know how rust future cancellation works!), it's possible we could rephrase this as:
`self.primary.run(req, workunit_store).select(delay.and_then(|()| command_runner.secondary.run(req_2, workunit_store_clone))` and then not need to do the matching ourselves... Not necessarily a recommendation, but a thought :)